server/mqtt.go

```diff
@@ -2833,6 +2959,30 @@ func (as *mqttAccountSessionManager) transferRetainedToPerKeySubjectStream(log *
 	return nil
 }
 
+func (as *mqttAccountSessionManager) getCachedRetainedMsg(subject string) *mqttRetainedMsg {
+	v, ok := as.rmsCache.Load(subject)
+	if !ok {
+		return nil
+	}
+	rm := v.(*mqttRetainedMsg)
+	if rm.expiresFromCache.Before(time.Now()) {
+		as.rmsCache.Delete(subject)
+		return nil
+	}
+	return rm
+}
+
+func (as *mqttAccountSessionManager) setCachedRetainedMsg(subject string, rm *mqttRetainedMsg, onlyReplace bool) {
```

Comment:
Now I see that in the case of `handleRetainedMsg()` we call this with `rm` that is created and reference `pp.msg` that is no longer copied, while in this case I believe it should (see comment about `pp` in that function), or at least make it a copy in `handleRetainedMsg()` when storing if `rm != nil`...

Reply:
Right, It was coded correctly without the wrappers before. Thanks for catching!